plugins/kubernetes/app/models/watchers/pod_watcher.rb

```diff
@@ -0,0 +1,84 @@
+module Watchers
+  class PodWatcher < BaseWatcher
+    POD_PHASES = %w[Pending Running Succeeded Failed Unknown]
+    
+    attr_reader :pod_info
+
+    def initialize(client, namespace, name: nil, label_selector: nil, log: true)
+      raise ArgumentError.new('Must specify either pod_name or label_selector') if name.blank? && label_selector.blank?
+      watcher = client.watch_pods(name: name, namespace: namespace, label_selector: label_selector)
+      super(watcher, log: log)
+
+      @pod_info = {}
+    end
+
+    def pod_names
+      @pod_info.keys
+    end
+
+    # Returns a hash with the number of pods in each status
+    def status_counts
+      counts = {}
+      POD_PHASES.each { |phase| counts[phase] = 0  }
+      @pod_info.each do |pod_name, notice|
+        counts[notice.phase] += 1
+      end
+      counts
+    end
+
+    def num_ready
+      @pod_info.values.count(&:ready?)
+    end
+
+    protected
+
+    # Notes on how to tell whether a deploy is successful or not:
+    #
+    # Keep track of the Pods that are created
+    # Watch for them to be all running and ready
+    # Once we have a set of all pod names, subscribe to events related to them
+    #   e.g. client.watch_events(fieldSelector: "involvedObject.name IN (pod1,pod2,...)")
+    #   look for any Events that look like a failure (e.g. failedScheduling)
+    # When all the pods are ready and running, mark the deploy as successful
+
+    def handle_notice(notice, &block)
+      return if handle_error(notice)
+      notice.extend(PWrapper)
```

Comment:
Runtime object extension used to be really expensive. I think it's been fixed in recent versions, but something to keep in mind if you're handling *a lot* of events.

http://dec0de.me/2013/02/dci-performance/
https://github.com/charliesome/charlie.bz/blob/master/posts/things-that-clear-rubys-method-cache.md